--- AppUIUtil.java.original	2018-09-10 15:35:30.997298300 +0200
+++ AppUIUtil.java	2018-09-11 20:16:52.177327900 +0200
@@ -36,6 +36,7 @@
 import com.intellij.openapi.util.Condition;
 import com.intellij.openapi.util.Pair;
 import com.intellij.openapi.util.SystemInfo;
+import com.intellij.openapi.util.io.FileUtil;
 import com.intellij.openapi.wm.ToolWindowManager;
 import com.intellij.ui.AppIcon.MacAppIcon;
 import com.intellij.ui.components.JBScrollPane;
@@ -60,9 +61,10 @@
 import java.awt.event.ActionEvent;
 import java.awt.event.AdjustmentEvent;
 import java.awt.event.AdjustmentListener;
-import java.io.File;
-import java.io.InputStream;
+import java.io.*;
+import java.lang.reflect.InvocationTargetException;
 import java.net.URL;
+import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
@@ -250,7 +252,36 @@
     return iconPath;
   }
 
+  public static void showBundleNoticeIfNeeded() {
+    String lib = new File(PathUtil.getJarPathForClass(AppUIUtil.class)).getParent();
+    File bundleAgreement = new File(lib + "/../data/config/options/bundleAgreement");
+
+    if (!bundleAgreement.exists()) {
+      try (Reader reader = new InputStreamReader(new BufferedInputStream(
+        AppUIUtil.class.getResourceAsStream("BundleAgreement.html")), StandardCharsets.UTF_8)) {
+
+        String text = new String(FileUtil.adaptiveLoadText(reader));
+        SwingUtilities.invokeAndWait(() -> showEndUserAgreementText("IntelliJ Scala Bundle Agreement", text));
+
+        File optionsDirectory = bundleAgreement.getParentFile();
+
+        if (!optionsDirectory.exists() && !optionsDirectory.mkdirs()) {
+          throw new IOException("Cannot create directory: " + bundleAgreement.getAbsolutePath());
+        }
+
+        try (BufferedOutputStream stream = new BufferedOutputStream(new FileOutputStream(bundleAgreement))) {
+          stream.write("accepted".getBytes(StandardCharsets.UTF_8));
+        }
+      }
+      catch (Exception e) {
+        Logger.getInstance(AppUIUtil.class).warn(e);
+      }
+    }
+  }
+
   public static void showUserAgreementAndConsentsIfNeeded() {
+    showBundleNoticeIfNeeded();
+
     if (ApplicationInfoImpl.getShadowInstance().isVendorJetBrains()) {
       EndUserAgreement.Document agreement = EndUserAgreement.getLatestDocument();
       if (!agreement.isAccepted()) {
@@ -301,6 +332,13 @@
    * @param isPrivacyPolicy  true if this document is a privacy policy
    */
   public static void showEndUserAgreementText(@NotNull String htmlText, final boolean isPrivacyPolicy) {
+    final String title = isPrivacyPolicy
+        ? ApplicationInfoImpl.getShadowInstance().getShortCompanyName() + " Privacy Policy"
+        : ApplicationNamesInfo.getInstance().getFullProductName() + " User License Agreement";
+    showEndUserAgreementText(title, htmlText);
+  }
+
+  private static void showEndUserAgreementText(String title, @NotNull String htmlText) {
     DialogWrapper dialog = new DialogWrapper(true) {
       @Override
       protected JComponent createCenterPanel() {
@@ -380,12 +418,7 @@
       }
     };
     dialog.setModal(true);
-    if (isPrivacyPolicy) {
-      dialog.setTitle(ApplicationInfoImpl.getShadowInstance().getShortCompanyName() + " Privacy Policy");
-    }
-    else {
-      dialog.setTitle(ApplicationNamesInfo.getInstance().getFullProductName() + " User License Agreement");
-    }
+    dialog.setTitle(title);
     dialog.setSize(JBUI.scale(509), JBUI.scale(395));
     dialog.show();
   }
