--- StartupUtil.java
+++ StartupUtil.java
@@ -114,6 +114,22 @@
       e.printStackTrace();
     }
 
+    if (SystemInfo.isMac && PathManager.getSystemPath().contains("/AppTranslocation/")) {
+      String message = "Please use Finder to move the application to another location before launching.\n\n" +
+                       "Since v10.12 macOS launches downloaded applications as read-only.\n" +
+                       "Moving application to any other location lifts that restriction.";
+      Main.showMessage("App Translocation detected", message, false);
+      System.exit(Main.DIR_CHECK_FAILED);
+    }
+
+    if (SystemInfo.isMac && !new File(PathManager.getSystemPath()).canWrite()) {
+      String message = "Please copy the application to your system before launching.\n\n" +
+                       "The application needs to write data, while DMG images are read-only.\n" +
+                       "Copying the program to Applications or to Desktop lifts the restriction.";
+      Main.showMessage("Read-only filesystem detected", message, false);
+      System.exit(Main.DIR_CHECK_FAILED);
+    }
+
     // note: uses config folder!
     if (!checkSystemFolders()) {
       System.exit(Main.DIR_CHECK_FAILED);
