--- AppUIUtil.java
+++ AppUIUtil.java
@@ -340,6 +340,13 @@
    * @param isPrivacyPolicy  true if this document is a privacy policy
    */
   public static void showEndUserAgreementText(@NotNull String htmlText, final boolean isPrivacyPolicy) {
+    final String title = isPrivacyPolicy
+        ? ApplicationInfoImpl.getShadowInstance().getShortCompanyName() + " Privacy Policy"
+        : ApplicationNamesInfo.getInstance().getFullProductName() + " User Agreement";
+    showEndUserAgreementText(title, htmlText, isPrivacyPolicy);
+  }
+
+  public static void showEndUserAgreementText(String title, @NotNull String htmlText, final boolean isPrivacyPolicy) {
     DialogWrapper dialog = new DialogWrapper(true) {
       private JEditorPane myViewer;
 
@@ -430,11 +437,7 @@
       }
     };
     dialog.setModal(true);
-    dialog.setTitle(
-      isPrivacyPolicy
-      ? ApplicationInfoImpl.getShadowInstance().getShortCompanyName() + " Privacy Policy"
-      : ApplicationNamesInfo.getInstance().getFullProductName() + " User Agreement"
-    );
+    dialog.setTitle(title);
     dialog.pack();
 
     SplashManager.executeWithHiddenSplash(dialog.getWindow(), () -> dialog.show());
